File: src/main/java/ttd/site/generator/YearMonth.java
package ttd.site.generator;

import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.springframework.util.Assert;

@Data
@RequiredArgsConstructor
class YearMonth implements Comparable<YearMonth> {

	private final int year, month;

	@Override
	public String toString() {
		return yearMonthKey(this);
	}

	private static String lpad(String theText, int numberOfCharsDesiredTotal, String characterToUseAsPadding) {
		Assert.hasText(theText, "the text to pad must be non-null");
		if (theText.length() < numberOfCharsDesiredTotal) {
			var delta = numberOfCharsDesiredTotal - theText.length();
			return characterToUseAsPadding.repeat(delta) + theText;
		}
		else {
			return theText;
		}
	}

	private static String yearMonthKey(YearMonth ym) {
		var y = ym.getYear();
		var m = ym.getMonth();
		var stringBuilder = new StringBuilder();
		var yearAsString = Integer.toString(y);
		var monthAsString = Integer.toString(m);
		stringBuilder.append(y < 10 ? lpad(yearAsString, 2, "0") : yearAsString);
		stringBuilder.append('-');
		stringBuilder.append(m < 10 ? lpad(monthAsString, 2, "0") : monthAsString);
		return stringBuilder.toString();
	}

	@Override
	public int compareTo(YearMonth o) {
		if (o instanceof YearMonth ym) {
			return ym.toString().compareTo(toString());
		}
		return -1;
	}

}

File: src/main/kotlin/ttd/site/generator/YearMonth.kt
package ttd.site.generator

import java.text.DateFormatSymbols
import java.util.*

class YearMonth(val year: Int, val month: Int) : Comparable<YearMonth> {

    override fun compareTo(other: YearMonth): Int = other.toString().compareTo(toString())

    private fun lpad(text: String, noOfChars: Int, paddingChar: String): String {
        return if (text.length < noOfChars) {
            val delta = noOfChars - text.length
            paddingChar.repeat(delta) + text
        } else {
            text
        }
    }

    private fun getMonthAsTextByLocale(month: Int, locale: Locale = Locale.US) =
        DateFormatSymbols.getInstance(locale).months[month - 1]


    fun toMonthString() = getMonthAsTextByLocale(this.month)

    fun toMonthCommaYearString() = "${getMonthAsTextByLocale(this.month)}, ${this.year}"


    override fun toString(): String {
        val y = this.year
        val m = this.month
        val stringBuilder = StringBuilder()
        val yearAsString = y.toString()
        val monthAsString = m.toString()
        stringBuilder.append(if (y < 10) lpad(yearAsString, 2, "0") else yearAsString)
        stringBuilder.append('-')
        stringBuilder.append(if (m < 10) lpad(monthAsString, 2, "0") else monthAsString)
        return stringBuilder.toString()
    }
}